### ISSUE TITLE
Softfail on optional dependencies

### ISSUE THREAD

**Issue by FFY00** [jaagr/polybar#1406]
## Is your feature request related to a problem? Please describe.
Yes, when polybar is built with an optional dependency, that optional dependency becomes a hard dependency. This is really bad for anyone trying to distribute binaries (ex. distros).

## Why does polybar need this feature?
So that polybar could be available distribution repos.

## Describe the solution you'd like
Instead of trying to blindly launch module, check if the dependencies are present first.

**Comment by patrick96** [comment]
Our current build setup links the shared libraries for the optional dependencies at compile time which makes the whole executable not start, if the libraries it was linked against are missing. The problem is not that we *blindly launch a module*, if the dependency is not present, the while `polybar` executable won't start.

There is a way to load shared libraries during runtime but it would be a major PITA to manually load every library function before we call it.

We had people having issues, because they removed an optional dependency after compilation, before so I'd also like to do this. However, I don't think there is a way to do this without either manually loading the library functions or major architectural changes.

@NBonaparte do you know of any other way to do this?

You can still package polybar for a distro, but you'll have to list the dependencies for all enabled features as hard dependencies.

**Comment by FFY00** [comment]
Why not changing to a plugin architecture? An object would be compiled for each module and then the main executable would try loading it. Your module structure seems relatively easy to port to something like that.

**Comment by patrick96** [comment]
That's what I meant when I was talking about major architectural changes. I don't think it would be that easy to port. We would need to create some way for polybar to interface with the modules and a way for modules to use helper functions inside polybar.

This is definitely something I'd like to see in the long run because it would make it easier to create more complex plugins and would also lessen the burden on us to create those plugins.

**Comment by skystrife** [comment]
I don't think it'll actually be that bad, but I could easily be underestimating the difficulty here.

I think the refactoring steps would be:

1. Update the CMake config to compile each module as a separate shared library, each of which links in whatever optional dependencies it requires. Then, make the `polybar` executable link in every (enabled) module's shared library.
2. Update [`make_module()`](https://github.com/jaagr/polybar/blob/master/include/modules/meta/factory.hpp#L55-L112) to use the "self-registering class" pattern (this will be a nice cleanup step anyway) where we instead look up a factory function from a singleton `map<string, std::function<module_interface*(const bar_settings&, string)>` rather than hard-coding the list of available modules. The basic idea here is to have some statically initialized class within each module's shared library whose constructor registers that module by name with the factory.
3. Update the CMake config to no longer link in each of the module libraries at compile time and instead `dlopen` them somewhere during the creation of the `controller`. On a successful `dlopen`, the module should be automagically registered with the factory by the statically initialized registration class; on a failure we can warn about a module not being supported, and `polybar` will gracefully exit with an error about an unknown module.

I can try to take a stab at this sometime this week.

**Comment by NBonaparte** [comment]
We currently list `xcb-util-{cursor, xrm}` as normal dependencies in the AUR packages, so we could probably just make them that way. These dependencies are pretty lightweight anyways so it shouldn't hurt that much.

@skystrife's method is the way I'd do it as well. `make_module()` is quite messy and is sorely in need of refactoring.

**Comment by FFY00** [comment]
Yeah, I am totally fine with the xcb dependencies. I would like to not make the user install i3 and all of it's dependencies just to run polybar though :/

Regarding the implementation, I understand it's not easy but once we have a working plugin model it's just refactoring the code, mainly, and overall it doesn't seem *that* hard.
